File: astar.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#define MOVES 4



struct Node {
	struct Node * pathParent;
	int arrival;
	int remaining;
	int row;
	int col;
	struct Node * left;
	struct Node * right;
	struct Node * graphParent;
};

struct PriorityQueue {
	struct Node * head;
};


struct Coord {
        int row;
        int col;
};

struct Tile {
	int type;
	int visited;
	struct Coord * pos;
};

struct NewMaze{
	struct Tile ** maze;
	struct Tile * start;
	struct Tile * end;
};


struct Node * NodeNew(struct Node * pathParent,int arrival,int remaining,int row,int col){
	struct Node * n = malloc(sizeof(struct Node ));
	n->pathParent = pathParent;
	n->arrival = arrival;
	n->remaining = remaining;
	n->row = row;
	n->col = col;
	n->left = NULL;
	n->right = NULL;
	n->graphParent = NULL;
	return n;
}

struct PriorityQueue * PQueueNew(struct Node * head){
	struct PriorityQueue * pq = malloc(sizeof(struct PriorityQueue));
	pq->head = head;
	return pq;
}

struct Coord * CoordNew(int row, int col){
        struct Coord * c = malloc(sizeof(struct Coord));
        c->row = row;
        c->col = col;
        return c;
}


struct Tile * TileNew(int type,int visited,int row,int col){
	struct Tile * t = malloc(sizeof(struct Tile));
	t->type = type;
	t->visited = visited;
	t->pos = CoordNew(row,col);
	return t;
}

struct NewMaze * NewMazeNew(struct Tile ** maze,struct Tile * start,struct Tile * end){
	struct NewMaze * n = malloc(sizeof(struct NewMaze));
	n->maze = maze;
	n->start = start;
	n->end = end;
	return n;
}


int estimate(struct Node * n){
	return n->arrival + n->remaining;
}

int matches(struct Node * n1,struct Node * n2){
	return n1->row == n2->row && n1->col == n2->col;
}

void insertNodeHelper(struct Node * parent,struct Node * toInsert){
	if (matches(parent,toInsert)){
        printf("HIT.\n");
		if (estimate(toInsert) < estimate(parent)){
			parent->pathParent = toInsert->pathParent;
			parent->arrival = toInsert->arrival;
		}
        return;
	}
	if (estimate(toInsert) >= estimate(parent)){
		if (!parent->right){
			toInsert->graphParent = parent;
			parent->right = toInsert;
			return;
		}
		insertNodeHelper(parent->right,toInsert);
	}else {
		if (!parent->left){
			toInsert->graphParent = parent;
			parent->left = toInsert;
			return;
		}
		insertNodeHelper(parent->left,toInsert);
	}
}


void insert(struct PriorityQueue * queue,struct Node * node){
	if (!queue->head){
		queue->head = node;
	}else{
		insertNodeHelper(queue->head,node);
	}
}

struct Node * takeClosestHelper(struct Node * parent){
	if (!parent->left){
		if (parent->right){
			struct Node * parentRoot = parent->graphParent;
			parentRoot->left = parent->right;
			parent->right->graphParent = parentRoot;
		}else{
			parent->graphParent->left = NULL;
		}

		return parent;
	}else{
		return takeClosestHelper(parent->left);
	}
}

struct Node * pop(struct PriorityQueue * queue){
	if (!queue->head){
		return NULL;
	}
	if (!queue->head->left){
		struct Node * toReturn = queue->head;
		if (queue->head->right){
			queue->head = queue->head->right;
		}else{
			queue->head = NULL;
		}
		return toReturn;
	}
	return takeClosestHelper(queue->head);
}


struct Tile * mazeAccess(struct Tile ** maze,int row,int col,int cols){
	struct Tile * t =  maze[row * cols + col];
    return t;
}

double rndm(){
	return ((double) rand() / (RAND_MAX));
}


int randRange(int min,int max){
	return rand() % (max + 1 - min) + min;
}

struct NewMaze * generateMaze(int width,int height,double freq){
	struct Tile ** maze = malloc ((width * height) * sizeof (struct Tile *));
	for (int i = 0;i<(width*height);i++){
		int row = i / width;
		int col = i % width;
		struct Tile * t = (rndm()<freq) ? TileNew(1,1,row,col) : TileNew(0,0,row,col);
		maze[i] = t;
	}
	struct Tile * startTile = mazeAccess(maze,randRange(0,height-1),randRange(0,width-1),width);
	struct Tile * endTile = mazeAccess(maze,randRange(0,height-1),randRange(0,width-1),width);
	startTile->type = 2;
	startTile->visited = 1;
	endTile->type = 3;
	return NewMazeNew(maze,startTile,endTile);

}

int square(int n){
	return n*n;
}

int pythagDistance(int row1,int row2,int col1,int col2){
	return square(row2-row1) + square(col2-col1);
}

//issues here
int tileGood(struct Tile ** maze,struct Tile * tile,int width,int height){
	return tile && tile->pos && (0 <= tile->pos->row) && (tile->pos->row < height) && (0 <= tile->pos->col) && (tile->pos->col < width) && !(mazeAccess(maze,tile->pos->row,tile->pos->col,width)->visited);
}

struct Coord * getCoordsForPair(struct Coord * pos,int rowMod,int colMod){
	return CoordNew(pos->row + rowMod,pos->col + colMod);
}

//may need to free tiles
struct Tile ** getAdjacentValidTiles(struct Tile ** maze,int height,int width,int row, int col){
	static int mods [MOVES * 2] = {0,-1,-1,0,0,1,1,0};
	struct Tile ** adjacent = malloc(MOVES * sizeof(struct Tile *));
	for (int i = 0;i<MOVES * 2;i+=2){
		struct Tile * t = mazeAccess(maze,row+mods[i],col+mods[i+1],width);
		if (tileGood(maze,t,width,height)){
			adjacent[i/2] = t;
		}else{
			adjacent[i/2] = NULL;
		}

	}
	return adjacent;
}

struct Coord ** unwrapPath(struct Node * end){
	struct Coord ** path = malloc((end->arrival + 1) * sizeof(struct Coord *));
	while (end){
		path[end->arrival] = CoordNew(end->row,end->col);
		end = end->pathParent;
	}
	return path;
}


struct Coord ** astar(struct NewMaze * maze,int height,int width){
    clock_t begin = clock();
	struct Coord * startPos = maze -> start -> pos;
	struct Coord * endPos = maze -> end -> pos;
    printf("Start: %d,%d\n",startPos->row,startPos->col);
    printf("End: %d,%d\n",endPos->row,endPos->col);
	struct Node * startNode = NodeNew(NULL,0,pythagDistance(startPos->row,endPos->row,startPos->col,endPos->col),startPos->row,startPos->col);
    struct PriorityQueue * queue = PQueueNew(startNode);
	while (queue->head){
		struct Node * position = pop(queue);
		if (position->row == endPos->row && position->col == endPos->col){
            clock_t term = clock();
            printf("%ld\n",(term - begin));
            printf("%ld\n",CLOCKS_PER_SEC);
			return unwrapPath(position);
		}
		struct Tile * current = mazeAccess(maze->maze,position->row,position->col,width);
        printf("%d\n",current->visited);
		current->visited = 1;
		struct Tile ** adjacent = getAdjacentValidTiles(maze->maze,height,width,position->row,position->col);
		for (int i = 0;i<MOVES;i++){
			if (adjacent[i]){
				struct Node * adj = NodeNew(position,position->arrival+1,pythagDistance(adjacent[i]->pos->row,endPos->row,adjacent[i]->pos->col,endPos->col),adjacent[i]->pos->row,adjacent[i]->pos->col);
				insert(queue,adj);
			}
		}
	}
    printf("FAILED\n");
	return NULL;
}



int main(){
	struct NewMaze * maze = generateMaze(100,100,0.1);
	astar(maze,100,100);
}
	

File: astar.go
package main

import (
	"math/rand"
	"time"
)

type TileType int

//Defines what the types of a maze tile can be.
const (
	FREE TileType = iota
	WALL
	START
	END
)

//Used for generating biotic maps, how many erosion cycles are made.
//Move to map gen package
const EROSIONS int = 10
//If a tile has less than MIN_TO_ERODE neighbors, it is cleared.
const MIN_TO_ERODE int= 2
//If a tile has more than MAX_TO_ERODE neighbors, it is filled.
const MAX_TO_ERODE int = 4

//If the path can include diagonal movements.
const DIAG bool = false

//The source of map generation randomness.
var source = rand.NewSource(time.Now().UnixNano())
//The random object used.
var random = rand.New(source)

//A holder for a map coordinate.
type Coord struct {
	Row int
	Col int
}

/**
A PriorityQueue Node that tracks a lot of state:
PathParent: The Node on the map that led to this Node.
Arrival: How many steps from the start Node it was to this Node.
Remaining: The heuristic estimate of how many more steps will be needed to reach the end Node.
Pos: The Coord position of the Node on the map.
Left: The Node to the left on the PriorityQueue tree.
Right: The Node to the right on the PriorityQueue tree.
GraphParent: The parent Node on the PriorityQueue tree of this Node.
 */
type Node struct {
	PathParent * Node
	Arrival int
	Remaining int
	Pos * Coord
	Left * Node
	Right * Node
	GraphParent * Node
}

//Just a wrapper for a Head Node.
type PriorityQueue struct {
	 Head * Node
}

/**
A representation of a map tile:
Pos: The coordinates of the Tile.
Type: The TileType (FREE,WALL,START,END) of the Tile.
Visited: If the Tile has already been visited.
 */
type Tile struct {
	Pos * Coord
	Type TileType
	Visited bool
}

//The possible moves that a path could take given diagonal movement.
//The first four are only lateral moves.
var terms8Axis = [8] * Coord{
	{
		Row: 0,
		Col: -1,
	},
	{
		Row: -1,
		Col: 0,
	},
	{
		Row: 0,
		Col: 1,
	},
	{
		Row: 1,
		Col: 0,
	},
	{
		Row: -1,
		Col: -1,
	},
	{
		Row: -1,
		Col: 1,
	},
	{
		Row: 1,
		Col: 1,
	},
	{
		Row: 1,
		Col: -1,
	},
}

//Estimates the remaining distance to the end from a given Tile.
func (n * Node) estimate()int{
	return square(n.Arrival) + n.Remaining
}

//Checks to see if two coordinates are in the same position.
func (c * Coord) matches (c1 * Coord) bool {
	return c.Row == c1.Row && c.Col == c1.Col
}

/**
Inserts a Tile Node into the PriorityQueue recursively.
parent: the Node to insert the new Node under.
node: the Node to insert into the PriorityQueue.
 */
func insertNodeHelper(parent * Node,node * Node){
	//If the parent is the child node:
	if parent.Pos.matches(node.Pos){
		//If the node was found in a faster way than last time:
		if node.Arrival < parent.Arrival{
			//Set the path that the parent came from to whatever the child came from.
			parent.PathParent = node.PathParent
			//Set the arrival time at the parent to the faster one of the child.
			parent.Arrival = node.Arrival
		}
		return
	}
	//If the child is farther from the end than parent:
	if node.estimate() >= parent.estimate(){
		//If the parent has no right Node, add the child.
		if parent.Right == nil{
			node.GraphParent = parent
			parent.Right = node
			return
		}
		//Else if it does, insert it under the right Node recursively.
		insertNodeHelper(parent.Right, node)
	//If the child is closer to the end than the parent:
	} else{
		//If the parent has no left Node, add the child.
		if parent.Left == nil{
			node.GraphParent = parent
			parent.Left = node
			return
		}
		//Else if it does, insert it under the left Node recursively.
		insertNodeHelper(parent.Left, node)
	}
}

/**
Inserts a Node into a PriorityQueue and handles the empty queue edge case.
 */
func (queue * PriorityQueue) insert(node * Node){
	//If the queue has nothing in it, set the head to the child Node.
	if queue.Head == nil {
		queue.Head = node
	//If the queue has a head, insert the child Node under the head.
	}else {
		insertNodeHelper(queue.Head, node)
	}
}

/**
Takes the Node with the shortest estimate under the parent.
 */
func takeClosestHelper(parent * Node) * Node {
	//If the parent has no left Node:
	if parent.Left == nil {
		//If the parent has a right Node, connects it to it's parent's left Node and removes itself.
		if parent.Right != nil {
			parentRoot := parent.GraphParent
			parentRoot.Left = parent.Right
			parent.Right.GraphParent = parentRoot
		//If the parent has no right node, just sets its own parent's left to nil and removes itself.
		}else{
			parent.GraphParent.Left = nil
		}
		return parent
	}
	//If left is not nil, recursively call this function again with the left node.
	return takeClosestHelper(parent.Left)
}


/**
Removes the Node with the least distance (arrival + estimated) to the end from a PriorityQueue.  Returns it.
Also handles some edge cases.
 */
func (queue * PriorityQueue) pop () * Node {
	//If the queue is empty, return nil
	if queue.Head == nil {
		return queue.Head
	}
	//If the queue has no left, return the head after setting the head's right as the new head.
	if queue.Head.Left == nil {
		toReturn := queue.Head
		queue.Head = queue.Head.Right
		return toReturn
	}
	//If the queue has a left, get the farthest left element recursively.
	return takeClosestHelper(queue.Head)
}

/**
Checks if a given tile is valid for maze exploration.
First checks if it is within the given bounds (x and y).
Next checks if the tile has already been visited.
 */
func tileGood(maze [][] * Tile, pos * Coord)bool{
	return 0 <= pos.Row && pos.Row < len(maze) && 0 <= pos.Col && pos.Col < len(maze[0]) && !maze[pos.Row][pos.Col].Visited
}

/**
Applies a coordinate shift to a pair of coordinates and returns a new Coord.

For example: (12,15),(1,0) returns (13,15).
 */
func getCoordsForPair(pos * Coord, mod * Coord)* Coord{
	return &Coord{
		Row: pos.Row + mod.Row,
		Col: pos.Col + mod.Col,
	}
}

/**
Applies simple biotic rules to soften edges of the maze.
Coordinates surrounded with mostly free space turn into free space.
Coordinates surrounded with mostly walls turn into walls.

Tiles bordering out of bounds spaces are counted as bordering walls there.
Doesn't validate for start/end, these would be added later.  Could check!
 */
func bioticErode(maze  [][] * Tile ){
	for row,wholeRow := range maze{
		for col := range wholeRow {
			surroundingWalls := 0
			for _,term := range terms8Axis{
				if !tileGood(maze,getCoordsForPair(&Coord{Row: row, Col: col},term)){
					surroundingWalls++
				}
			}
			if surroundingWalls > MAX_TO_ERODE {
				maze[row][col].Type = WALL
				maze[row][col].Visited = true
			}else if surroundingWalls < MIN_TO_ERODE {
				maze[row][col].Type = FREE
				maze[row][col].Visited = false
			}
		}
	}
}

/**
Randomly generates a maze of a certain size with a wall frequency.
Returns the maze as well as the start and end positions.
 */
func generateMaze(width int,height int,freq float64)([][] * Tile,*Coord,*Coord) {
	maze := make([][] * Tile,height)
	for i := 0;i<height;i++{
		maze[i] = make([] * Tile,width)
		for b := 0;b<width;b++{
			if random.Float64() < freq{
				maze[i][b] = &Tile{
					Pos:     &Coord{
						Row: i,
						Col: b,
					},
					Type:    WALL,
					Visited: true,
				}
			}else{
				maze[i][b] = &Tile{
					Pos:     &Coord{
						Row: i,
						Col: b,
					},
					Type:    FREE,
					Visited: false,
				}
			}
		}
	}

	for i := 0;i<EROSIONS;i++{
		bioticErode(maze)
	}

	start := &Coord{
		Row: random.Intn(height-1),
		Col: random.Intn(width - 1),
	}
	maze[start.Row][start.Col].Type = START
	maze[start.Row][start.Col].Visited = false

	end := &Coord{
		Row: random.Intn(height - 1),
		Col: random.Intn(width - 1),
	}
	maze[end.Row][end.Col].Type = END
	maze[end.Row][end.Col].Visited = false

	return maze,start,end
}

func (t * Terminal) toString()string{
	str := ""
	for _,row := range t.DataHistory {
		sub := make([]byte,t.Width)
		//s1 := ""
		for b,col := range row {
			sub[b] = col.top().BackgroundCode
			//s1 += string(int(col[t.Depth - 1].BackgroundCode))
		}
		str += string(sub) + "\n"
	}
	return str
}

/**
Reads the terminal space and returns a parsed map.
Takes the symbol that represents a wall, the start, and the end.
Returns a maze array (2d) and the start/end coordinates.
 */
func (t * Terminal) parseMazeFromCurrent(wall byte, start byte, end byte) ([][]*Tile, *Coord, *Coord) {
	//The dimensions of the terminals memory
	height := len(t.DataHistory)
	width := len(t.DataHistory[0]) - 1
	maze := make([][] * Tile, height)
	//Create default start and end coords
	startCoord := &Coord{
		Row: 0,
		Col: 0,
	}
	endCoord := &Coord{
		Row: 0,
		Col: 0,
	}
	discoveredWalls := 0
	//Reads through each tile in the array and switches on the background code to set the maze
	for i := 0;i<height;i++{
		row := make([] * Tile, width)
		maze[i] = row
		for b := 0;b<width;b++{
			//Default is free space.
			maze[i][b] = &Tile{
				Pos:     &Coord{
					Row: i,
					Col: b,
				},
				Type:    FREE,
				Visited: false,
			}
			switch t.DataHistory[i][b].top().BackgroundCode {
			case wall:
				maze[i][b].Type = WALL
				maze[i][b].Visited = true
				discoveredWalls++
				break
			case start:
				maze[i][b].Type = START
				maze[i][b].Visited = false
				startCoord.Row = i
				startCoord.Col = b
				break
			case end:
				maze[i][b].Type = END
				maze[i][b].Visited = false
				endCoord.Row = i
				endCoord.Col = b
				break
			}
		}
	}
	return maze,startCoord,endCoord
}

/**
Very similar to above function, but it takes the array of characters as an argument.  Might refactor and combine these two.
 */
func parseMazeFromChars(data [][]rune, wall rune, start rune, end rune) ([][]*Tile, *Coord, *Coord) {
	height := len(data)
	width := len(data[0])
	maze := make([][] * Tile, height)
	startCoord := &Coord{
		Row: 0,
		Col: 0,
	}

	endCoord := &Coord{
		Row: 0,
		Col: 0,
	}
	for i := 0;i<height;i++{
		row := make([] * Tile, width)
		maze[i] = row
		for b := 0;b<width;b++{
			maze[i][b] = &Tile{
				Pos:     &Coord{
					Row: i,
					Col: b,
				},
				Type:    FREE,
				Visited: false,
			}
				switch data[i][b] {
				case wall:
					maze[i][b].Type = WALL
					maze[i][b].Visited = true
					break
				case start:
					maze[i][b].Type = START
					maze[i][b].Visited = false
					startCoord.Row = i
					startCoord.Col = b
					break
				case end:
					maze[i][b].Type = END
					maze[i][b].Visited = false
					endCoord.Row = i
					endCoord.Col = b
					break
				}
		}
	}
	return maze,startCoord,endCoord
}

//Quickly squares two numbers.
func square(n int)int{
	return n*n
}

/**
Finds the non square rooted distance between two Coordinates.
 */
func pythagDistance(c1 * Coord,c2 * Coord)int{
	return square(c1.Row - c2.Row) + square(c1.Col - c2.Col)
}

/**
For a given maze and coordinate, gets an array of all adjacent coordinates.
If a coordinate is invalid, fills its array slot with nil.
 */
func getAdjacentValidTiles(maze [][] * Tile,pos * Coord) [] * Coord  {
	resultSize := 4
	if DIAG{
		resultSize = 8
	}
	adjacents := make([] * Coord,resultSize)
	for i,term := range terms8Axis[0:resultSize]{
		newCoord := getCoordsForPair(pos,term)
		if tileGood(maze,newCoord){
			adjacents[i] = newCoord
		}else{
			adjacents[i] = nil
		}
	}
	return adjacents
}

/**
Given an end Node with a path parent leading backwards down the correct path,
returns an array of Nodes composing the path.
 */
func unwrapPath(end * Node)  [] * Coord {
	path := make([] * Coord,end.Arrival + 1)
	for end != nil {
		path[end.Arrival] = &Coord{
			Row: end.Pos.Row,
			Col: end.Pos.Col,
		}
		end = end.PathParent
	}
	return path
}

/**
Performs the A* pathfinding algorithm on a maze of Tiles.  Returns the path as an array.
 */
func astar(maze [][] * Tile,start * Coord,end * Coord) [] * Coord {
	//Creates the start node.
	s := &Node{
		PathParent:  nil,
		Arrival:     0,
		Remaining:   pythagDistance(start,end),
		Pos:         start,
		Left:        nil,
		Right:       nil,
		GraphParent: nil,
	}
	//Creates the PriorityQueue with the start Node as the head.
	queue := PriorityQueue{Head: s}
	//Defines a variable to track current position.
	var position * Node
	//Defines a variable to track the current maze data.
	var mazeData * Tile
	//Creates a reused array of coords the size of maxAdj.
	var adjacent [] * Coord
	//While the queue has Nodes in it:
	for queue.Head != nil {
		//Get the closest node as position.
		position = queue.pop()
		//Get the details about that node as mazeData.
		mazeData = maze[position.Pos.Row][position.Pos.Col]
		//If this Node has already been explored, skip it and let it die.
		//This happens when two tiles adjacent to this one have both been considered.
		if mazeData.Visited{
			continue
		}
		//If the new Node is the end Node:
		if position.Pos.matches(end){
			//Return the path that landed at the end Node.
			return unwrapPath(position)
		}
		//Marks the tile as visited.
		mazeData.Visited = true
		//Gets all adjacent spots to the current position.
		adjacent = getAdjacentValidTiles(maze,position.Pos)
		for _,tile := range adjacent{
			if tile != nil{
				//For all valid adjacent Nodes, insert them into the PriorityQueue.
				queue.insert(&Node{
					PathParent:  position,
					Arrival:     position.Arrival + 1,
					Remaining:   pythagDistance(tile,end),
					Pos:         tile,
					Left:        nil,
					Right:       nil,
					GraphParent: nil,
				})
			}
		}
	}
	return nil
}
